NFTMarketplace.sol

```diff
@@ -141,70 +184,60 @@ contract NFTMarketplace is ReentrancyGuard, Ownable {
-                activeCount++;
-            }
-        }
+    // Function to get marketplace statistics such as total volume, average price, and total listings (live)
+    function getMarketplaceStatisticsLive() external view returns (MarketStatistics memory data) {
+        // Calculate average price if there are sold listings
+        uint256 averagePrice = totalListings > 0 ? totalVolume / totalListings : 0;
 
-        // Create an array to hold the active listings
-        Listing[] memory activeListings = new Listing[](activeCount);
-        uint256 currentIndex = 0;
-        // Populate the array with active listings
-        for (uint256 i = 0; i < listings.length; i++) {
-            if (listings[i].active && listings[i].tokenAddress == _tokenAddress) {
-                activeListings[currentIndex] = listings[i];
-                currentIndex++;
-            }
-        }
-        return activeListings;
+        return data = MarketStatistics(
+            totalVolume,
+            averagePrice,
+            totalListings
+        );
     }
 
-    // Function to get marketplace statistics such as total volume, average price, and total listings
-    function getMarketplaceStatistics() public view returns (uint256 totalVolume, uint256 averagePrice, uint256 totalListings) {
-        uint256 totalPrice = 0;
-        uint256 totalSold = 0;
-        // Loop through all listings to calculate statistics
-        for (uint256 i = 0; i < listings.length; i++) {
-            if (!listings[i].active) { // Assuming a listing is inactive once sold
-                totalSold++;
-                totalPrice += listings[i].price;
-            }
-        }
-        // Calculate average price if there are sold listings
-        averagePrice = totalSold > 0 ? totalPrice / totalSold : 0;
-        // Total listings is the length of the listings array
-        totalListings = listings.length;
-        // Total volume is the sum of all sold listing prices
-        totalVolume = totalPrice;
-        return (totalVolume, averagePrice, totalListings);
+    // Function to get marketplace statistics such as total volume sold, average price, and total listings (sold)
+    function getMarketplaceStatisticsSold() external view returns (MarketStatistics memory data) {
+        uint256 averagePrice = totalSold > 0 ? totalVolumeSold / totalSold : 0;
+
+        return data = MarketStatistics(
+            totalVolumeSold,
+            averagePrice,
+            totalSold
+        );
     }
 
-    // Function to get all active listings by a specific seller
-    function getActiveListingsBySeller(address _seller) public view returns (Listing[] memory) {
-        // Count active listings for the seller
-        uint256 activeCount = 0;
-        for (uint256 i = 0; i < listings.length; i++) {
-            if (listings[i].active && listings[i].seller == _seller) {
-                activeCount++;
+    // Function to get all active listings for a specific NFT contract
```

Comment:
the loop here and in getListingsByContract() should be removed, perhaps there will come a time when the total number of players will be greater than the maximum gas per block